src/book/03-guides/01-main-codebase/02-nuclear-tutorial.mdx

```diff
@@ -0,0 +1,312 @@
+
+Up at the top of the file, include the protobuf message we just created. Note that these messages are compiled into C++ code, so include the compiled header.
+
+```cpp
+#include "message/PingPong.hpp"
+```
+
+Make sure it compiles and still runs successfully. Because we added files, we will need to configure again.
+
+```bash
+./b configure
+./b build
+./b run testprogram
+```
+
+If this works, you can move onto the next section where we make our reactors.
+
+### Reactors
+
+Reactors are chunks of code that will run given certain conditions are true. We will make two reactors, one that runs reactions whenever it gets a `Pong` message, and the other when it gets a `Ping` message.
+
+Head back to `TestModule.cpp`. Underneath the Configuration reactor, add in the following reactor
+
+```cpp
+on<Trigger<Ping>>().then([this](const Ping& ping) {
+    // Some code!
+});
+```
+
+`Trigger` means that it will run when it a `Ping` message is available. We get access to this message via the `ping` parameter. Let's add in the other reactor.
+
+```cpp
+on<Trigger<Pong>>().then([this](const Pong& pong) {
+    // Some code!
+});
+```
+
+Very similar, but it gets the `Pong` message instead.
+
+Next, we want to get the count from the `Ping` or `Pong` message, increment it, and send a new message with the incremented count. Here's the code for the first reactor.
+
+```cpp
+// Print the Ping message!
+log<NUClear::INFO>("Ping count", ping.count);
+
+// Make a Pong message to send
+auto pong = std::make_unique<Pong>();
+pong->count = ping.count + config.increment;
+
+// Send the message
+emit(pong);
+```
+
+And for the second reactor
+
+```cpp
+// Print the Pong message!
+log<NUClear::INFO>("Pong count", pong.count);
+
+// Make a Pong message to send
+auto ping = std::make_unique<Ping>();
+ping->count = pong.count + config.increment;
+
+// Send the message
+emit(ping);
+```
+
+Have a look through the comments and make sure you understand the code. `auto` is a keyword that automatically determines the type of the variable.
+
+Is everything ready now to see the messages ping-ponging between each other, printing out increasing count values? Compile the program and run it, and see what happens.
+
+```cpp
+./b build
+./b run testprogram
+```
+
+Well... Nothing changed. If you're not sure why, have a think about it before moving onto the next section.
+
+### Starting the Chain
+
+The problem is, that there is no initial Ping or Pong message. No one ever sends the first message - the first reactor is forever waiting for a Ping message, and the second reactor is forever waiting for a Pong message!
+
+Ok, so how do we start the reaction? There are a few ways it could be done, but we will use the `Startup` NUClear DSL. This is another reactor that will run only when the program starts up.
```

Comment:
```suggestion
Ok, so how do we start the reaction tasks? There are a few ways it could be done, but we will use the `Startup` NUClear domain specific language (DSL) word. This is another reaction that will run only when the program starts up.
```